### ISSUE TITLE
strange inventory behavior with inventory directory

### ISSUE THREAD

**Issue by bitmensch** [ansible/ansible#13081]
```
find . -type f | xargs -n1 -t cat
cat ./inventory/item1
hostname1

cat ./inventory/item2
hostname1 ansible_connection=docker ansible_host=b0271bf10b34

cat ./playbook.yml

---
- hosts: hostname1
  tasks:
    - ping:

ansible-playbook -i inventory/ playbook.yml

PLAY ***************************************************************************

TASK [setup] *******************************************************************
fatal: [hostname1]: UNREACHABLE! => {"changed": false, "msg": "ERROR! SSH encountered an unknown error during the connection. We recommend you re-run the command using -vvvv, which will enable SSH debugging output to help diagnose the issue", "unreachable": true}

PLAY RECAP *********************************************************************
hostname1                  : ok=0    changed=0    unreachable=1    failed=0
```

Switching content of `inventory/item1` and `inventory/item2` makes things work again:

```
ansible-playbook -i inventory/ playbook.yml

PLAY ***************************************************************************

TASK [setup] *******************************************************************
ok: [hostname1]

TASK [ping] ********************************************************************
ok: [hostname1]

PLAY RECAP *********************************************************************
hostname1                  : ok=2    changed=0    unreachable=0    failed=0
```

I would expect things to work without any specific order.

**Comment by Yannig** [comment]
Thanks for this description but the problem is also present with previous devel version ...

Previously, you were talking about a dynamic script and here, we have 2 statics files. Maybe, the problem occurs using a dynamic inventory.

**Comment by bitmensch** [comment]
@Yannig: Yes, you're right. I was looking at the wrong place.

**Comment by bitmensch** [comment]
Anyway, I would consider this another bug. I updated the description to be no longer specific to a particular commit.

**Comment by jimi-c** [comment]
This is not quite a bug, though it's something I've felt we should raise an error over. Host names in Ansible are considered unique keys, and inventory directories process files in alphabetical order. So the first file is processed and the host is added, however when it finds the same host in the second, it checks to see if the host is already present and silently ignores it because it does in fact already exist. The silent ignore is what I believe should instead be an error, or at least a warning.

**Comment by itldo** [comment]
I'm encountering similar strange behavior with a combination of a static hosts file and a dynamic ec2 inventory. I use ec2 instance names as ansible host keys, and provide an ansible_ssh_host (and an ansible_host) variable along with the dynamic inventory.

Then I use the ec2 names in groups defined in the static hosts file.
Before stumbling on this Issue, I suspected an ordering problem and I found that naming my dynamic inventory file to sort ahead of the static file solved some, but not all of the problems. It seemed to just shift the problems to different hosts. In some cases the ansible_host is used, in other cases the host key is used for connecting. Also the behavior can be inconsistent between a playbook and ad-hoc ansible, for the same host and the same inventory.

Rather than issuing an error or a warning, I think the assignment of inventory variables needs to be consistent and well defined. This is a showstopper issue for me, because I cannot touch my inventory (e.g. create a new group and add a host) without randomly breaking things.
I'll try to come up with a reproducible example.
Thanks.

**Comment by amenonsen** [comment]
@itldo did you manage to write up an example I could try?

**Comment by itldo** [comment]
@amenonsen I dug into this yesterday trying to understand the problem using the inventories that were causing my problem. Had to work on something else today, so I'll try to pick this up again tomorrow. Thanks.

**Comment by itldo** [comment]
@amenonsen I have some automated failing homegrown testcases in the form of a few scripts. What is a good way to get this to you. I can also post the output, which prints its inventory fixtures, to this or another thread.

**Comment by itldo** [comment]
@amenonsen I'm sure that was the stupidest question ever asked on github.
I've created a repo with the test cases at 

  http://github.com/itldo/ansible_test_var

**Comment by jimi-c** [comment]
Hi all, this appears to have been the same bug as #13397, which was fixed by f467f17. 

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:

https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code
Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
 @jimi-c
